src/msg/async/ProtocolV2.cc

```diff
@@ -463,16 +463,26 @@ void ProtocolV2::send_message(Message *m) {
                    << dendl;
     if (((!replacing && can_write) || state == STANDBY) && !write_in_progress) {
       write_in_progress = true;
+
+      /* unlock the mutex now because dispatch_event_external() may
+         block waiting for another mutex */
+      l.unlock();
```

Comment:
Is it possible to race on the deallocation of `write_handler` with `AsyncConnection::_stop()` -> `AsyncConnection::cleanup()`?

Reply:
Maybe there (already) is, but it has nothing to do with this PR; `cleanup()` doesn't hold `write_lock`.
(Ceph's shutdown is a big mess anyway, see https://github.com/ceph/ceph/pull/60320 and related PRs.)

Reply:
Agreed, it looks `dispatch_event_external()` always conveys through the `external_events` queue, so this should be quite isolated. Thanks.